wait.PollUntilContextCancel immediately executes condition once

AxeZhan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
In #118686, we make sure that wait.PollUntilContextTimeout function, if immediate is true, the condition will be invoked before waiting and guarantees that the condition is invoked at least once, regardless of whether the context has been cancelled. However, the changes in that pr will make new PollUntilContextCancel function immediately executes given condition twice, which is not mentioned anywhere. Which is not semantically same with wait.PollImmediateUntil.
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #119533

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
wait.PollUntilContextTimeout function, if immediate is true, the condition will be invoked before waiting and guarantees that the condition is invoked at least once and then wait a interval before executing again.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

AxeZhan commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -203,45 +203,49 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
+			ctx, cancel := contextFn()
+			defer cancel()
+
+			timer := Backoff{Duration: time.Millisecond}.Timer()
```
I increased this a little bit(from Microsecond to Millisecond), because the new test case use a timeout to control the context, it can be very flaking when using a millisecond(Probably due to precision issues in golang). I think it's still fine because this is still a very fast test. But I can also separate this test case if it's needed.

AxeZhan replied:
Alright, `Millisecond` runs fine on my local machine. But it's still flaking in the ci, so I separated a `Second` version.

AxeZhan commented:
/cc @aojea

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -203,45 +203,49 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
 	}
 
 	for _, test := range tests {
-		for _, immediate := range []bool{true, false} {
```
use separete commits for these things please, so it it easier to review, so you don't add a new test, remove some dead code and modify the existing test in the same commit

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop.go:
```diff
@@ -50,13 +50,13 @@ func loopConditionUntilContext(ctx context.Context, t Timer, immediate, sliding
-		}
+	}
+
+	timeCh = t.C()
```
@smarterclayton the problem I see know is how to handle the sliding in this first iteration

aojea replied:
but this change looks correct to me, just that `if !sliding` this block should be executed always before the `immediate` function and `if sliding` always after the `immediate` function

aojea replied:
oh, nevermind, immediate guarantees the execution, so this changes seems correct as it is now

smarterclayton replied:
Hrm, I don't think this preserves sliding properly for the first iteration.

If sliding is false, then the sliding interval should start (invocation of `t.C()`) before the condition is executed (even in immediate).  If sliding is true, the sliding interval should start after the condition is executed.  Where timeCh is initialized matters (that's why it's lazily initialized in the loop).

The fact that a test case didn't fail means that we aren't properly testing for that condition.  Please add a test case that ensures sliding isn't broken in this scenario (which should then make it clear what the fix should be)?

aojea replied:
/lgtm cancel
@AxeZhan please take a look at Clayton´s comment

aojea commented:
/assign @aojea @smarterclayton

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -206,42 +206,92 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
+	}
+}
+
+func TestLoopConditionUntilContextSemanticTime(t *testing.T) {
```
do we need to add a new test, can not add just as a new test case in one of the existing tests?

AxeZhan replied:
At first I created a new test, because only this case requires a larger time durtation. But now I think we can make this configurable in the test :), I add this  as a new case in the new commit.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -203,45 +206,54 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
 			errExpected:      context.DeadlineExceeded,
 		},
+		{
+			name:      "context canceld before the second execution",
```
```suggestion
			name:      "context canceled before the second execution and immediate",
```

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -203,45 +206,54 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
+
+			timer := test.timer
+			if timer == nil {
+				timer = defaultTimer
```
I think that you have to keep creating a new timer in each iteration and not reuse the same to avoid pollution between testscases

aojea replied:
```suggestion
				timer = Backoff{Duration: time.Microsecond}.Timer()
```

AxeZhan replied:
Ah yes. Updated.

aojea commented:
@AxeZhan ideally we'll have to backport this, so check if is feasible to have the fix + the new testcases in the same commit ( the dead code removal can be done in a subsequent commit since does not affect the result)

AxeZhan commented:
@aojea Updated, now the first commit is code changes(including main code and new test code), the second one is removal of the dead code.

AxeZhan commented:
/retest

aojea commented:
/lgtm

Thanks

@smarterclayton PTAL

alexzielenski commented:
/triage accepted

liangyuanpeng commented:
Just for confirm that, it does mean the `wait.PollUntilContextCancel immediately`  have a wrong behavior? Thanks.

liangyuanpeng commented:
Thanks for your clarify.

aojea commented:
> we're having some issues with preserves sliding properly for the first iteration.

I think that this sliding problem is a bug we have in this and the previous implementation

AxeZhan commented:
I've updated the code and added a test case for sliding.
The logic of this case goes like this:

1. create a `loopConditionUntilContext`with a timer(one second each iteration), and a condition function(cost one second each time ), also with a context that will be canceld in 5 seconds.
2.  now if sliding is false,  then period includes the runtime for condition. So we'll never wait between two calls of the condition. Thus, we expected condition function to be called for 4 times.
3.  if sliding is true,  then period doesn't include the runtime for condition. So basically we will need two seconds for each iteration(one second to run contdition, and one second to wait the timer). And condition function should be called for 3 times.(It will finish the first two times, and get canceld when sleeping in third iteration).

I don't know if this case can ensures sliding isn't broken , I'd like to hear your opinions :)

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -203,45 +204,89 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
+			immediate: true,
+			sliding:   true,
+			context: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 5*time.Second)
```
5 seconds is very expensive, I think we can have the same functionality this way

```
diff --git a/staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go b/staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go
index af8f0c9a721..fa984cc0d5a 100644
--- a/staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go
@@ -218,11 +218,11 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
                        timer:            Backoff{Duration: 2 * time.Second}.Timer(),
                },
                {
-                       name:      "long duration of condition and sliding false",
+                       name:      "inmidiate and long duration of condition and sliding false",
                        immediate: true,
                        sliding:   false,
                        context: func() (context.Context, context.CancelFunc) {
-                               return context.WithTimeout(context.Background(), 5*time.Second)
+                               return context.WithTimeout(context.Background(), time.Second)
                        },
                        callback: func(attempts int) (bool, error) {
                                if attempts >= 4 {
@@ -231,15 +231,16 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
                                time.Sleep(time.Second)
                                return false, nil
                        },
-                       attemptsExpected: 4,
-                       timer:            Backoff{Duration: time.Second, Jitter: 0.001}.Timer(),
+                       errExpected:      context.DeadlineExceeded,
+                       attemptsExpected: 2,
+                       timer:            Backoff{Duration: 100 * time.Millisecond}.Timer(),
                },
                {
-                       name:      "long duration of condition and sliding true",
+                       name:      "inmidiate and long duration of condition and sliding true",
                        immediate: true,
                        sliding:   true,
                        context: func() (context.Context, context.CancelFunc) {
-                               return context.WithTimeout(context.Background(), 5*time.Second)
+                               return context.WithTimeout(context.Background(), time.Second)
                        },
                        callback: func(attempts int) (bool, error) {
                                if attempts >= 4 {
@@ -249,8 +250,8 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
                                return false, nil
                        },
                        errExpected:      context.DeadlineExceeded,
-                       attemptsExpected: 3,
-                       timer:            Backoff{Duration: time.Second, Jitter: 0.001}.Timer(),
+                       attemptsExpected: 1,
+                       timer:            Backoff{Duration: 100 * time.Millisecond}.Timer(),
                },
        }
```

AxeZhan replied:
5 seconds is indeed expensive, I reduced the duration in each test to 20% of the original, and it seems works fine.
But I don't think we can remove the `Jitter` here, cause if so, we'll create a `fixedTimer` which doesn't implement the `Next` right?
```
func (t *fixedTimer) Next() {
	// no-op for fixed timers
}
```

aojea commented:
/lgtm
/assign @smarterclayton 

```
stress ./wait.test -test.run loop -test.v
5s: 48 runs so far, 0 failures
10s: 144 runs so far, 0 failures
15s: 240 runs so far, 0 failures
20s: 336 runs so far, 0 failures
ç25s: 432 runs so far, 0 failures
```

AxeZhan commented:
Another thing: I think we don't need the lazy check in the loop anymore because it will be initialized before the loop anyway, right?

aojea commented:
> Another thing: I think we don't need the lazy check in the loop anymore because it will be initialized before the loop anyway, right?

yeah, but we want to get this merged and backported, so these things can be follow up PRs

/lgtm

AxeZhan commented:
/cc @smarterclayton

aojea commented:
/assign @liggitt @wojtek-t

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -203,45 +204,89 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
 			attemptsExpected: 0,
 			errExpected:      context.DeadlineExceeded,
 		},
+		{
```
From the coverage profile it looks like we're missing coverage in one scenario:

> sliding=true, condition fails one time

Can you add tests for 

1. sliding=true, condition fails once, then passes
2. sliding=true, condition fails once, then has a context cancellation error

smarterclayton replied:
Ok, with the right version checked out coverage shows:

```
        for {
                // checking ctx.Err() is slightly faster than checking a select
                if err := ctx.Err(); err != nil {
                        return err // uncovered
                }
```

This now immediately follows

```
	select {
	case <-doneCh:
		return ctx.Err()
	case <-timeCh:
	}
```

and is redundant (before, we could conditionally check).  We also check this before the loop.  So the only condition this `if` block can be called in now is when we wait the full duration of `<-timeCh` and then in the nanoseconds between that check and condition the context is cancelled.  That's not necessary, so we should remove the first if block in the main `for {}` loop.

AxeZhan replied:
> That's not necessary, so we should remove the first if block in the main for {} loop.
Updated.

AxeZhan replied:
Remove this caused some flaky tests:
```
{Failed;Failed;Failed;  === RUN   Test_loopConditionUntilContext_semantic/context_cancelled_after_5_attempts
    loop_test.go:286: expected attempts count: 5 but got: 6
    --- FAIL: Test_loopConditionUntilContext_semantic/context_cancelled_after_5_attempts (0.00s)
;=== RUN   Test_loopConditionUntilContext_semantic/context_cancelled_and_immediate_after_5_attempts
    loop_test.go:286: expected attempts count: 5 but got: 6
    --- FAIL: Test_loopConditionUntilContext_semantic/context_cancelled_and_immediate_after_5_attempts (0.00s)
;=== RUN   Test_loopConditionUntilContext_semantic
--- FAIL: Test_loopConditionUntilContext_semantic (2.62s)
;}
```
Same as before, I can't reproduce these failures on my mac. I assume it's because the removal of this check. Let me test with the CI.

AxeZhan replied:
@smarterclayton This removal seems can break some existing uts, so I increased the timer's duration in the tests(see the newest commit).

aojea replied:
> That's not necessary, so we should remove the first if block in the main `for {}` loop.

@smarterclayton remember that was redundant but was there on purpose, see the comment on line 85. we should keep it

```
		// NOTE: b/c there is no priority selection in golang
		// it is possible for this to race, meaning we could
		// trigger t.C and doneCh, and t.C select falls through.
		// In order to mitigate we re-check doneCh at the beginning
		// of every loop to guarantee at-most one extra execution
		// of condition.
```

smarterclayton replied:
Hrm.  If I set a really really fast timer, it's a good point that we could never execute.  However, now that we've eliminated the conditional invocation, we should move it up in the loop:

```
	if sliding {
		timeCh = t.C()
	}

	for {
	        // Wait for either the context to be cancelled or the next invocation be called
		select {
		case <-doneCh:
			return ctx.Err()
		case <-timeCh:
		}
		// IMPORTANT: Because there is no channel priority selection in golang
		// it is possible for very short timers to "win" the race in the previous select
		// repeatedly even when the context has been canceled.  We therefore must
		// explicitly check for context cancellation on every loop and exit if true to
		// guarantee that we don't invoke condition more than once after context has
		// been cancelled.
		if err := ctx.Err(); err != nil {
			return err
		}
		
		if !sliding {
			t.Next()
		}
		if ok, err := func() (bool, error) {
			defer runtime.HandleCrash()
			return condition(ctx)
		}(); err != nil || ok {
			return err
		}
		if sliding {
			t.Next()
		}
	}
```

I think this is more correct now (and i reworded the comments to capture that) - please try this out with your new tests.

AxeZhan replied:
@smarterclayton  Updated, and I think you're right. Now we can pass all the tests without increase the timer's duration.

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -203,45 +204,89 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
+				if attempts >= 4 {
+					return true, nil
+				}
+				time.Sleep(time.Second / 5)
```
Are you trying to cause it to race?

AxeZhan replied:
> Are you trying to cause it to race?

I explained my thoughts in [comment](https://github.com/kubernetes/kubernetes/pull/119762#issuecomment-1671778866). 

Basically, I created a callback function which requires a long duration to execute. So in a same duration, poll with sliding true/false will execute this function different times.  

If sliding is false, we'll keep execute this function and never waits, because timech is drained during the sleep in condition function. Otherwise, we'll execute less times because we'll have to wait a longer time before next execution.

AxeZhan replied:
> How did you choose this amount to sleep?

Well, at first it was a big amount, and aojea suggests to reduce the time cost. So I tried to reduce it, but if this amount is too small, this test will fail in the CI(however it passed on my mac). I doubt it should be some accuracy problem. So I tried a few amounts, and this amount is ok on both sides(CI and my mac), and 1 second seems not too expensive.

aojea replied:
it took me a while to understand it too, I had to draw it actually, it plays with the periods and the attempts to make it predictable

smarterclayton commented:
Some more comments to address.

AxeZhan committed:
fix wait.PollUntilContextCancel immediately executes condition once

AxeZhan committed:
remove dead code in test

AxeZhan committed:
add new test cases

smarterclayton commented:
/approve
/lgtm

aojea commented:
/milestone v1.29

this is a bugfix
